### ISSUE TITLE
[support] stage 2 training with CP>1

### ISSUE THREAD

**Issue by berserkr** [NVIDIA-NeMo/Megatron-Bridge#3258]
### User problem

Hi, it is my understanding that calculate_per_token_loss: false is required for sample-level loss... Stage 2 in the paper, however, it seems that this requires CP==1? Does this mean that Nemotron at 512k was trained without CP?

If the above is not correct, would it be possible to tell me exactly how stage 2 was done with 512k context length? Just EP alone is enough?

### Desired outcome

Support for calculate_per_token_loss: false when CP>1

### Affected area

area:model

### Urgency / use case

Blocking current work

**Comment by berserkr** [comment]
Follow up... so I realized that there was a fix sometime back for CP>1: https://github.com/NVIDIA-NeMo/Megatron-Bridge/issues/2914 --- the torch.where fix in losses.py protects against the NaN-on-all-masked-rank issue @cuichenx  - maybe that fix you applied many be needed after all

**Comment by yaoyu-33** [comment]
Just to confirm, this is a nemotron reproduce question. If you are not aware, we have a repro recipe provided here https://github.com/NVIDIA-NeMo/Nemotron/blob/main/docs/nemotron/super3/README.md

**Comment by cuichenx** [comment]
by stage 2 do you mean SFT or CPT?

**Comment by berserkr** [comment]
@cuichenx SFT :)  --- @yaoyu-33 --- trying to reproduce stage 2 (sample level).

**Comment by cuichenx** [comment]
I see, I'll consult expert on this and get back to you

**Comment by cuichenx** [comment]
Sorry still checking with the experts. Seems like you found a real issue with the documentation :)

**Comment by cuichenx** [comment]
Hi, turns out the sample-level average is not the same as the `calculate_per_token_loss` argument. In an unreleased nemotron training branch, there is an option to to set an alpha exponent in a two-stage per-sample loss normalization for packed-sequence SFT

Two stages:
  1. Per-sample reweight — divide each sample's loss mask by clamp(token_count^alpha, min=1). After this, sample i sums to token_count^(1-alpha).
  2. Per-pack rescale — rescale the whole pack so total weight equals num_samples_in_pack (= beta * real_samples + (1-beta) * iter_samples/GBS),
  preserving global normalization.

 In SFT stage 2, calculate_per_token_loss=True and alpha = 1.0, so that every sample contributes equally regardless of length.
Let me know if this answers your question.

**Comment by berserkr** [comment]
@cuichenx are there plans to release that branch? This is my understanding: 
```
  Stage 1 is straightforward:
  - Each sample's loss_mask divided by token_count^alpha
  - Result: sample i total weight = token_count / token_count^alpha = token_count^(1-alpha)
  - With alpha=1: every sample sums to 1
  
  Stage 2 is about what happens across different packs:

  Different packs have different numbers of samples packed in:
  - Pack A: 5 short samples (total weight after stage 1 = 5)
  - Pack B: 1 long sample (total weight after stage 1 = 1)
  
  Without rescaling, Pack A gets 5x the gradient influence. Stage 2 rescales each pack's total weight to:

  num_samples_in_pack = beta * real_samples + (1-beta) * iter_samples/GBS

  Where:    
  - real_samples = actual number of samples in THIS pack (e.g., 5 for Pack A, 1 for Pack B)
  - iter_samples = total samples across the entire global batch 
  - GBS = number of packs in the global batch
  - iter_samples/GBS = average samples per pack
  
  What beta controls:

  ┌──────┬─────────────────────────────┬────────────────────────────────────────────────────────────┐
  │ beta │     Each pack's weight      │                           Effect                           │
  ├──────┼─────────────────────────────┼────────────────────────────────────────────────────────────┤
  │ 1.0  │ real_samples                │ True sample-level — each sample globally gets equal weight │
  ├──────┼─────────────────────────────┼────────────────────────────────────────────────────────────┤
  │ 0.0  │ iter_samples/GBS (constant) │ Every pack gets same weight regardless of sample count     │
  └──────┴─────────────────────────────┴────────────────────────────────────────────────────────────┘
  

  Total weight across all packs:
  sum_j (beta * real_samples_j + (1-beta) * iter_samples/GBS)
  = beta * sum(real_samples_j) + (1-beta) * GBS * (iter_samples/GBS)
  = beta * iter_samples + (1-beta) * iter_samples 
  = iter_samples  ← always the same, regardless of beta
```

**Comment by cuichenx** [comment]
Your understanding is essentially correct — the two-stage decomposition (per-sample by alpha, per-pack by beta) and the algebra both match the
  implementation. Two small notes:

  1. The stage-1 divisor in the code is actually torch.clamp(token_count ** alpha, min=1) — the min=1 floor guards against samples with very few
  unmasked tokens, so the "sample weight = token_count^(1-alpha)" identity only strictly holds when token_count^alpha >= 1. Conceptually it doesn't
  change anything.
  2. Worth flagging that the "beta=1.0 → each sample globally equal weight" reading is true given alpha=1.0 too. Alpha and beta are independent
  knobs:
    - alpha controls within-pack distribution (alpha=1 → samples equal regardless of length; alpha=0 → token averaging)
    - beta controls cross-pack distribution (beta=1 → pack weight = its sample count; beta=0 → every pack gets equal total weight)

  With alpha<1, beta=1 normalizes the pack total to real_samples but longer samples within a pack still dominate. Pure "every sample equal globally"
   needs alpha=1 + beta=1 — which is what Krishna's launch script uses.

  As for releasing the branch — unfortunately we don't have plans to upstream that code right now, sorry about that. Happy to keep chatting through
  the design if it's useful for what you're building.
